x-pack/plugins/fleet/server/services/epm/packages/_install_package.ts

```diff
@@ -146,17 +146,45 @@ export async function _installPackage({
       installMlModel(packageInfo, paths, esClient, savedObjectsClient, logger, esReferences)
     );
 
-    // installs versionized pipelines without removing currently installed ones
-    esReferences = await withPackageSpan('Install ingest pipelines', () =>
-      installPipelines(packageInfo, paths, esClient, savedObjectsClient, logger, esReferences)
+    /**
+     * In order to install assets in parallel, we need to split the preparation step from the installation step. This
+     * allows us to know which asset references are going to be installed so that we can save them on the packages
+     * SO before installation begins. In the case of a failure during installing any individual asset, we'll have the
+     * references necessary to remove any assets in that were successfully installed during the rollback phase.
+     *
+     * This split of prepare/install could be extended to all asset types. Besides performance, it also allows us to
+     * more easily write unit tests against the asset generation code without needing to mock ES responses.
+     */
```

Comment:
@kpollich Here would be a good place to start reviewing the latest commit / refactor

Reply:
Thanks - super helpful comment here so much appreciated on that.